lib/FreeRTOS/freertos-plus-mqtt/agent_command_pool.c

```diff
@@ -0,0 +1,136 @@
+ * by either an error or the execution of the commands callback.
+ */
+static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
+
+/**
+ * @brief A counting semaphore used to guard the pool of Command_t structures.  To
+ * obtain a structure first decrement the semaphore count.  To return a structure
+ * increment the semaphore count after the structure is back in the pool.
+ */
+static SemaphoreHandle_t freeCommandStructMutex = NULL;
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t blockTimeMs )
+{
+    Command_t * structToUse = NULL;
+    size_t i;
+    static bool initialized = false;
+
+    if( !initialized )
+    {
+        memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+        freeCommandStructMutex = xSemaphoreCreateCounting( MQTT_COMMAND_CONTEXTS_POOL_SIZE, MQTT_COMMAND_CONTEXTS_POOL_SIZE );
+        configASSERT( freeCommandStructMutex ); /*_RB_ Create all objects here statically. */
+
+        initialized = true;
+    }
+
+    /* Check counting semaphore has been created. */
+    if( freeCommandStructMutex != NULL )
+    {
+        /* If the semaphore count is not zero then a command context is available. */
+        if( xSemaphoreTake( freeCommandStructMutex, pdMS_TO_TICKS( blockTimeMs ) ) == pdPASS )
+        {
+            for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+            {
+                taskENTER_CRITICAL();
+                {
+                    /* If the commandType is NONE then the structure is not in use. */
+                    if( commandStructurePool[ i ].commandType == NONE )
+                    {
+                        LogDebug( ( "Removed Command Context %d from pool", ( int ) i ) );
+                        structToUse = &( commandStructurePool[ i ] );
+
+                        /* To show the struct is no longer available to be returned
+                         * by calls to Agent_ReleaseCommand(). */
+                        structToUse->commandType = !NONE;
+                        taskEXIT_CRITICAL();
+                        break;
+                    }
+                }
+                taskEXIT_CRITICAL();
+            }
+        }
+    }
+
+    return structToUse;
+}
+
+/*-----------------------------------------------------------*/
+
+bool Agent_ReleaseCommand( Command_t * pCommandToRelease )
+{
+    size_t i;
+    bool structReturned = false;
+
+    /* See if the structure being returned is actually from the pool. */
+    for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+    {
+        if( pCommandToRelease == &( commandStructurePool[ i ] ) )
+        {
+            /* Yes its from the pool.  Clearing it to zero not only removes the old
+             * data it also sets the structure's commandType parameter to NONE to
+             * mark the structure as free again. */
+            memset( ( void * ) pCommandToRelease, 0x00, sizeof( Command_t ) );
```

Comment:
Doesn't this need to be with in taskENTER_CRITICAL() and taskEXIT_CRITICAL() as this modifies the array?

Reply:
No it does not, since it is checking if the addresses match, and the point of `Agent_GetCommand()` is to ensure that the same address is not used more than once at a time.